### PR TITLE
Always set a message-id on mails

### PR DESCRIPTION
**Pull request by Garionion**
Solves #17899
This PR makes sure, that every Mail contains a "Message-ID" header.
the ID has the format "YYYYMMDDHHMMSS.<md5hash-of-the-message>@domain"
i used an md5 hash here because it is faster than sha256 and i thougth that because the id also contains a timestamp, the likelyhood of an id-collision is not very high.

### CODE REVIEW

File: services/mailer/mailer.go [hunk 1]
```diff
@@ -75,6 +83,18 @@ func (m *Message) SetHeader(field string, value ...string) {
...
 
+func (m *Message) generateMessageID() string {
+	var b bytes.Buffer
+	err := gob.NewEncoder(&b).Encode(m)
+	if err != nil {
+		log.Warn("Error converting mail message to bytes, falling back to randomness: %v", err)
+		uid, _ := uuid.NewRandom()
+		b.WriteString(uid.String())
+	}
+	now := time.Now().Format("20060102150405") // YYYYMMDDHHMMSS
```
Comment (wxiaoguang) [inline services/mailer/mailer.go]: I think we do not need these complex methods. a `UUID` or `TimeMillSec + (4~8)RandomBytes` is enough.

File: services/mailer/mailer.go [hunk 2]
```diff
@@ -67,6 +70,11 @@ func (m *Message) ToMessage() *gomail.Message {
 		msg.SetBody("text/plain", plainBody)
 		msg.AddAlternative("text/html", m.Body)
 	}
+
+	if len(msg.GetHeader("Message-ID")) == 0 {
+		msg.SetHeader("Message-ID", "<"+m.generateMessageID()+">")
+		log.Warn("Mail to %s with subject %s did not contain a Message-Id, I set one for you.")
```
Comment (wxiaoguang) [inline services/mailer/mailer.go]: These warning doesn't seem to help users (or developers)

If we want to indicate that the message id is auto-generated, we can make the message-id like: `autogen-{UUID}@domain`

File: services/mailer/mailer.go (hunk 1 above)
Comment (Garionion) [reply services/mailer/mailer.go]: but isn't randomness slower than just use the message itself?

File: services/mailer/mailer.go (hunk 2 above)
Comment (Garionion) [reply services/mailer/mailer.go]: ok, i prefixed the generated id with `autogen` :)

File: services/mailer/mailer.go [hunk 3]
```diff
@@ -67,6 +70,10 @@ func (m *Message) ToMessage() *gomail.Message {
 		msg.SetBody("text/plain", plainBody)
 		msg.AddAlternative("text/html", m.Body)
 	}
+
+	if len(msg.GetHeader("Message-ID")) == 0 {
+		msg.SetHeader("Message-ID", "<autogen-"+m.generateMessageID()+">")
```
Comment (KN4CK3R) [inline services/mailer/mailer.go]: The result is `<autogen-<...@...>>` which does not look correct.

File: services/mailer/mailer.go (hunk 1 above)
Comment (KN4CK3R) [reply services/mailer/mailer.go]: How about `TimeMilSec + (len(m.To[0]) ^ len(m.Subject) ^ len(m.Body))`. That should be unique enough per message and does not need random numbers or hash calculation.

Comment (Garionion) [reply services/mailer/mailer.go]: seems good, changed to that

File: services/mailer/mailer.go (hunk 3 above)
Comment (Garionion) [reply services/mailer/mailer.go]: oh, you're right

**Comment by mscherer** [comment]
@Garionion you need to run "make fmt" to reformat the code so the CI pass

**Comment by wxiaoguang** [comment]
`UnixMilli` is available since go 1.17, Gitea still supports go 1.16.

**Comment by Garionion** [comment]
yeah i just noticed -.-

**Comment by wxiaoguang** [comment]
I added unit tests for this function.

**Review by wxiaoguang — APPROVED** [review]

**Review by lunny — APPROVED** [review]

**Review by wxiaoguang — APPROVED** [review]

**Pull request merged**